### PR TITLE
fix(storage, ios): resolve listAll promise once and only once on error

### PR DESCRIPTION
**Pull request by TPXP**
### Description

When calling `listAll` on a forbidden directory, the promise is rejected then RNFirebase tries to resolve it with an empty result, which leads to the following Exception (which I believe would crash the app in production):
```
Invariant Violation: No callback found with cbID 256 and callID 128 for  RNFBStorageModule.listAll - most likely the callback was already invoked. Args: '[{"nextPageToken":null,"items":[],"prefixes":[]}]'
2020-12-18 11:45:24.318162+0100 ExampleApp[40550:189312] [native] Running surface LogBox ({
    initialProps =     {
    };
    rootTag = 11;
})
```
The root cause is the lack of a `return`on this line of the Firebase iOS SDK : https://github.com/firebase/firebase-ios-sdk/blob/14764b8d60a6ad023d8fa5b7f81d42378d92e6fe/FirebaseStorage/Sources/FIRStorageReference.m#L417

This issue does not occur on Android

### Related issues
Upstream issue: https://github.com/firebase/firebase-ios-sdk/issues/7197

### Release Summary

* [Storage] FIX: Avoid a crash if listAll fails

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [ ] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

Try it on your project !
<details>
<summary>Storage rules</summary>

```
rules_version = '2';
service firebase.storage {
  match /b/{bucket}/o {
    match /stuff/{userId} {
      allow read: if false;
    }
  }
}
```
</details>

<details>
<summary>JS Code</summary>

```
import storage from "@react-native-firebase/storage";
storage().ref('stuff').listAll(console.log, console.warn);
```
</details>

The app will crash without this change on iOS.

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Comment by mikehardy** [comment]
Whoa! Very interesting, Do you have an upstream issue logged with the sdk team? I'm okay in principle with workarounds and will review, but only if upstream is aware and I didn't see an issue ref. If I missed it sorry

**Comment by TPXP** [comment]
I didn't check with the upstream team, it might be relevant to raise an issue on their side although it makes sense to return the items the query was able to fetch as well as the error (which is something we can't do with promises).

In any case, I didn't touch the Android code, this PR aligns behaviors on both platforms (taking Android as a reference).

File: packages/storage/e2e/StorageReference.e2e.js [hunk 1]
```diff
@@ -346,6 +346,20 @@ describe('storage() -> StorageReference', function() {
       result.prefixes.length.should.be.greaterThan(0);
       result.prefixes[0].constructor.name.should.eql('StorageReference');
     });
+
+    it('should not crash if the user is not allowed to list the directory', async function() {
```
Comment (mikehardy) [inline packages/storage/e2e/StorageReference.e2e.js:350]: thank you so much for adding a test to support this change!

File: packages/storage/ios/RNFBStorage/RNFBStorageModule.m [hunk 2]
```diff
@@ -194,7 +194,16 @@ - (void)invalidate {
 ) {
   FIRStorageReference *storageReference = [self getReferenceFromUrl:url app:firebaseApp];
 
+  __block bool alreadyCompleted = false;
+
   id completionBlock = ^(FIRStorageListResult *result, NSError *error) {
+    // This may be called multiple times if an error occurs
+    // Make sure we won't try to resolve the promise twice in this case
+    // See https://github.com/firebase/firebase-ios-sdk/blob/14764b8d60a6ad023d8fa5b7f81d42378d92e6fe/FirebaseStorage/Sources/FIRStorageReference.m#L417
```
Comment (mikehardy) [inline packages/storage/ios/RNFBStorage/RNFBStorageModule.m:202]:
```suggestion
    // TODO - remove pending resolution of https://github.com/firebase/firebase-ios-sdk/issues/7197
```

File: packages/storage/ios/RNFBStorage/RNFBStorageModule.m [hunk 3]
```diff
@@ -194,7 +194,16 @@ - (void)invalidate {
...
+    if(alreadyCompleted) {
```
Comment (mikehardy) [inline packages/storage/ios/RNFBStorage/RNFBStorageModule.m:203]:
```suggestion
    if (alreadyCompleted) {
```

**Review by mikehardy — APPROVED** [review]
Great! I subscribed to the upstream issue and replaced the code reference in your workaround with the issue link. I also fixed a blank space nit

Crashing in native code is never acceptable so I'm positive on a workaround now that it is being tracked upstream, and this looks like a good workaround.

The test coverage is wonderful, I personally appreciate that as it gives me confidence when doing releases, thank you.

**Comment by mikehardy** [comment]
Android E2E is having issues but they are unrelated, and I'm tracking those separately, this is good to merge

**Comment by mikehardy** [comment]
@TPXP deploying as v10.3.1 as I type this, cheers

**Comment by TPXP** [comment]
Wow, that was fast. Thanks for your responsiveness! <3

**Comment by mikehardy** [comment]
All the work over the last few months and years of developing test coverage, the last few months stabilizing the CI tests and polishing the release system means it is now very easy to do one of my favorite things: crush crashes :-). Always my pleasure to help crush a native crash. Thanks for the PR.

**Comment by mikehardy** [comment]
Upstream just closed with a PR, tagged for firebase-ios-sdk 7.4.0

**Pull request merged**